### PR TITLE
Inconsistent Wildcard handling for JPA

### LINKED ISSUE #20610: Inconsistent Wildcard handling for JPA
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

storage

### Describe the bug

As a follow-up on #20510, we found that when searching for users, JPA allows wildcards `%' and `_` when searching, while other providers like LDAP allow only `*`. As `*` is a natural wildcard character in LDAP, it works in all places, while with JPA it only works at the beginning and the end of the search string. 

### Version

21.1.1

### Expected behavior

The only wildcard characters should be `*` which should work consistently across all providers in all places in the search string. All special characters in a specific provider like `%` and `_` for JPA should be escaped.

### Actual behavior

An asterisk work with JPA only the beginning and end of the string, but not in the middle. Native JPA wildcard work in the search string anywhere in the string.

### How to Reproduce?

Use 

```
./kcadm.sh get users -q "search=Wilson"
```

with 

* `Wil*son` -> finds entries with wildcard in LDAP, but not in JPA, as JPA only replaces asterisks at the beginning and the end, and `*` is a natural wildcard in LDAP.
* `Wil%son` -> finds entries with wildcard in JPA, but not LDAP, as `%` is a natural wildcard in JPA

### Anything else?

LDAP was handled to support `*` as expected in #20510

### PR DESCRIPTION
**Pull request by vramik**
Closes #20610

### CODE REVIEW

**Review by ghost — COMMENTED** [review]
Unreported flaky test detected, please review

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java [hunk 1]
```diff
@@ -897,7 +897,7 @@ protected void ensureEmailConstraint(List<UserEntity> users, RealmModel realm) {
     }
 
     private Predicate[] getSearchOptionPredicateArray(String value, CriteriaBuilder builder, From<?, UserEntity> from) {
-        value = value.toLowerCase();
+        value = value.toLowerCase().replace("%", "\\%").replace("_", "\\_");
```
Comment (ahus1) [inline model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java:900]: I gave it a thought I think first all existing `\` should be escaped first before escaping the special characters.

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java [hunk 2]
```diff
@@ -910,16 +910,8 @@ private Predicate[] getSearchOptionPredicateArray(String value, CriteriaBuilder
...
-                if (value.length() > 0 && value.charAt(value.length() - 1) == '*') {
-                    value = value.substring(0, value.length() - 1);
-                }
-                value += "%";
-            }
+            value = value.replace("*", "%");
+            if (value.isEmpty() || value.charAt(value.length() - 1) != '%') value += "%";
 
             orPredicates.add(builder.like(from.get(USERNAME), value));
             orPredicates.add(builder.like(from.get(EMAIL), value));
```
Comment (ahus1) [inline model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java:917]: I had a look and one can add a thirst parameter to `.like()` with specifying the escape character. I had a look around what other frameworks do, and they seem to have `\` as the default escape character. Still there seems to be some intrinsic behavior at least in MS SQL where one can use backets to escape parts of a `LIKE`, unless you explicitly state the escaping character - see https://www.mytecbits.com/microsoft/sql-server/escape-square-brackets-in-like.

So I'd suggest to use the third parameter to make the escaping explicit in the code, and safe from MS SQL intrinsic. As `\` is then used in multiple places, it IMHO calls for a constant to be defined in this class for this value.

**Review by ahus1 — CHANGES_REQUESTED** [review]
Thank you for providing the PR. This will make the behavior more consistent.

See below for some additional comments. 

As Keycloak's search behavior is something very central, I suggest to add this change to the release notes / migration guide as well. When mentioning the wildcard search with `*`, it would be helpful for readers to point out that adding quotes `"w*ord"` will do an exact search (as before).

**Review by ahus1 — COMMENTED** [review]
@vramik - I reviewed the code and the tests worked nicely. I did some additional manual testing and found a corner case: 

When using an exact search like `"us%er"`, there shouldn't be any escaping. So I added one more commit where this is then moved into the LIKE block.

Please have a look if this change is OK from your point of view. Thanks!

**Comment by vramik** [comment]
> Please have a look if this change is OK from your point of view. Thanks!

LGTM, thank you for the fix!

**Review by ahus1 — APPROVED** [review]
Thank you for this change which makes the behavior a lot more consistent for the users!

**Pull request merged**